### ISSUE TITLE
Intermittent issue during Patching Update.exe icon

### ISSUE THREAD

**Issue by chrisrestall** [clowd/Clowd.Squirrel#72]
When running the squirrel command for generating the exe and nupkg, I periodically (about 1/20) times get this issue

```
INFO] DotnetUtil: Extracting Update.exe resources to temp directory
[INFO] DotnetUtil: Patching Update.exe icon

[ERRO] System.AggregateException: One or more errors occurred. (Failed to modify resources, command invoked was: 'C:\tools\Clowd_squirrel\rcedit.exe System.String[]'

Output was:
Fatal error: Unable to commit changes)
 ---> System.Exception: Failed to modify resources, command invoked was: 'C:\tools\Clowd_squirrel\rcedit.exe System.String[]'

Output was:
Fatal error: Unable to commit changes
   at Squirrel.HelperExe.SetExeIcon(String exePath, String iconPath) in ./Internal/HelperExe.cs:line 131
   at SquirrelCli.DotnetUtil.UpdateSingleFileBundleIcon(String sourceFile, String destinationFile, String iconPath) in ./DotnetUtil.cs:line 142
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean )
   at System.Threading.Tasks.Task.Wait(Int32 , CancellationToken )
   at SquirrelCli.Program.Releasify(ReleasifyOptions options) in ./Program.cs:line 179
   at SquirrelCli.Program.Pack(PackOptions options) in ./Program.cs:line 149
   at SquirrelCli.CommandAction`1.Execute(IEnumerable`1 args) in ./ValidatedOptionSet.cs:line 160
   at SquirrelCli.CommandSet.Execute(String[] args) in ./ValidatedOptionSet.cs:line 195
   at SquirrelCli.Program.Main(String[] args) in ./Program.cs:line 87

Squirrel 2.8.40+1c583f, tool for creating and deploying Squirrel releases
```
It typically goes away when I rerun.

**Comment by peppy** [comment]
Are the files you are operating on potentially on some kind of network/cloud storage?

**Comment by caesay** [comment]
Probably related to https://github.com/clowd/Clowd.Squirrel/issues/46. The issue there was updating PE resources on network-attached storage, and the solution was just to add retries to Squirrel around the `Kernel32.UpdateResource` call. rcedit.exe uses the same windows function. 

The proper fix for this is probably to remove our dependency on rcedit.exe and add some code to `ResourceUpdater.cs` allowing us to patch an icon without modifying other resources.

**Comment by chrisrestall** [comment]
This is running on a local filesystem directory.

**Comment by caesay** [comment]
This is also related: https://github.com/dotnet/runtime/issues/3832. There, they found that some people's AntiVirus systems were locking the file that was being updated, and disabling the antivirus fixed the intermittent failures. (not that I am suggesting doing this, just that it's a possible cause)

I am leaving this issue open for now, until I have time to drop rcedit in favor of doing the icon updating internally.

**Comment by chrisrestall** [comment]
Our company uses an anti-malware tool called Carbon-black which might be the culprit.  The when is indeterminant.  I've asked for a scan exception so we'll see if that makes any difference.

**Comment by caesay** [comment]
Closing because this is fixed in the `cs/xplat` branch. I am not going to backport this fix to 2.x but it will be released in 3.0.

**Comment by chrisrestall** [comment]
Thanks for the fix and the awesome work done on the fork!


Sent from my iPhone

On May 7, 2022, at 8:16 AM, Caelan Sayler ***@***.***> wrote:

﻿

Closing because this is fixed in the cs/xplat branch. I am not going to backport this fix to 2.x but it will be released in 3.0.
